src/googleclouddebugger/firebase_client.py

```diff
@@ -0,0 +1,587 @@
+    ref = firebase_admin.db.reference(path)
+    self._breakpoint_subscription = ref.listen(self._ActiveBreakpointCallback)
+
+  def _ActiveBreakpointCallback(self, event):
+    if event.event_type == 'put':
+      if event.data is None:
+        # Either deleting a breakpoint or initializing with no breakpoints.
+        # Initializing with no breakpoints is a no-op.
+        # If deleting, event.path will be /{breakpointid}
+        if event.path != '/':
+          breakpoint_id = event.path[1:]
+          del self._breakpoints[breakpoint_id]
+      else:
+        if event.path == '/':
+          # New set of breakpoints.
+          self._breakpoints = {}
+          for (key, value) in event.data.items():
+            self._AddBreakpoint(key, value)
+        else:
+          # New breakpoint.
+          breakpoint_id = event.path[1:]
+          self._AddBreakpoint(breakpoint_id, event.data)
+
+    elif event.event_type == 'patch':
+      # New breakpoint or breakpoints.
+      for (key, value) in event.data.items():
+        self._AddBreakpoint(key, value)
+    else:
+      native.LogWarning('Unexpected event from Firebase: '
+                        f'{event.event_type} {event.path} {event.data}')
+      return
+
+    native.LogInfo(f'Breakpoints list changed, {len(self._breakpoints)} active')
+    self.on_active_breakpoints_changed(list(self._breakpoints.values()))
+
+  def _AddBreakpoint(self, breakpoint_id, breakpoint_data):
+    breakpoint_data['id'] = breakpoint_id
+    self._breakpoints[breakpoint_id] = breakpoint_data
+
+  def _TransmitBreakpointUpdates(self):
+    """Tries to send pending breakpoint updates to the backend.
+
+    Sends all the pending breakpoint updates. In case of transient failures,
+    the breakpoint is inserted back to the top of the queue. Application
+    failures are not retried (for example updating breakpoint in a final
+    state).
+
+    Each pending breakpoint maintains a retry counter. After repeated transient
+    failures the breakpoint is discarded and dropped from the queue.
+
+    Args:
+      service: client to use for API calls
+
+    Returns:
+      (reconnect, timeout) tuple. The first element ("reconnect") is set to
+      true on unexpected HTTP responses. The caller should discard the HTTP
+      connection and create a new one. The second element ("timeout") is
+      set to None if all pending breakpoints were sent successfully. Otherwise
+      returns time interval in seconds to stall before retrying.
+    """
+    retry_list = []
+
+    # There is only one consumer, so two step pop is safe.
+    while self._transmission_queue:
+      breakpoint_data, retry_count = self._transmission_queue.popleft()
+
+      bp_id = breakpoint_data['id']
+
+      try:
+        # Something has changed on the breakpoint.
+        # It should be going from active to final, but let's make sure.
+        if not breakpoint_data['isFinalState']:
+          raise BaseException(
```

Comment:
That's a fair question.  You're right; this is catching a programming error.  That being said, the exception is being caught below and will be logged as "a fatal error" and the breakpoint update will not be retried.  Functionally I believe that this is reasonable but the usage of an exception to accomplish it may still be in question.

Thoughts?

Reply:
Leaving as is per offline discussion.